### ISSUE TITLE
Research how to get all the different tokens in a ERC-721

### ISSUE THREAD

**Issue by piersss** [airswap/airswap-marketplace#22]
This is for filipmanole

**Comment by piersss** [comment]
I asked chat GPT a few questions about this. Maybe it will help you.

**Comment by piersss** [comment]
To use OpenZeppelin's ERC721 contract to get all the owners of a token, you will need to do the following:

    Install the OpenZeppelin Contracts library. You can do this using npm by running the following command:

npm install @openzeppelin/contracts

    Import the ERC721 contract from the OpenZeppelin Contracts library. You can do this by adding the following line to the top of your code file:

const { ERC721 } = require('@openzeppelin/contracts');

    Instantiate the ERC721 contract by passing in the contract address and the ABI (Application Binary Interface) of the contract. You can find the ABI of the contract by going to Etherscan and clicking on the "Contract" tab.

const erc721Contract = new ERC721(contractAddress, contractAbi);

    Call the getOwners function on the ERC721 contract instance to get an array of all the token holders.

const owners = await erc721Contract.methods.getOwners().call();

Keep in mind that this function may not return a complete list of all the token holders, as some token holders may have privacy concerns and may not want their address to be publicly visible on the Ethereum blockchain.

You will also need to make sure that you have a web3 provider set up and connected to the Ethereum network that the contract is deployed on. You can use a tool like Web3.js or a wallet application like MetaMask to do this.

**Comment by piersss** [comment]
Using OpenZeppelin's ERC721 contract, you can get a list of all the tokens of an ERC721 token that are listed for sale by calling the tokenOfOwnerByIndex function on the contract instance. This function takes two arguments: the address of the owner and the index of the token in the owner's token array. It returns the token ID of the token at the specified index.

To get a list of all the tokens that are listed for sale, you will need to do the following:

    Install the OpenZeppelin Contracts library. You can do this using npm by running the following command:

npm install @openzeppelin/contracts

    Import the ERC721 contract from the OpenZeppelin Contracts library. You can do this by adding the following line to the top of your code file:

const { ERC721 } = require('@openzeppelin/contracts');

    Instantiate the ERC721 contract by passing in the contract address and the ABI (Application Binary Interface) of the contract. You can find the ABI of the contract by going to Etherscan and clicking on the "Contract" tab.

const erc721Contract = new ERC721(contractAddress, contractAbi);

    Call the balanceOf function on the ERC721 contract instance to get the number of tokens that a particular address holds.

const balance = await erc721Contract.methods.balanceOf(address).call();

    Iterate through the token indices and call the tokenOfOwnerByIndex function for each index to get the token ID of the token at that index.

for (let i = 0; i < balance; i++) {
  const tokenId = await erc721Contract.methods.tokenOfOwnerByIndex(address, i).call();
  console.log(tokenId);
}

This will give you a list of all the token IDs of the tokens that the specified address holds. You can then use this information to find out more about the tokens and see if they are listed for sale on a marketplace or exchange.

Keep in mind that this approach will only work for tokens that are owned by a single address. If the tokens are owned by multiple addresses, you will need to iterate over all the addresses and get their token lists in the same way.

You will also need to make sure that you have a web3 provider set up and connected to the Ethereum network that the contract is deployed on. You can use a tool like Web3.js or a wallet application like MetaMask to do this.

**Comment by piersss** [comment]
Right now we're looking into this: https://stackoverflow.com/questions/69302924/erc-721-how-to-get-all-token-ids

**Comment by gpxl-dev** [comment]
@piersss & @filipmanole, the above stack overflow answer is basically what I was talking about on the call.

"Mints" are implemented by transferring an NFT from the zero address (i.e. `0x0000000000000000000000000000000000000000`) to the receiver (similarly burns are the inverse). The transfer event is emitted each time with the to & from in the topics.

The stack overflow is also correct in that it's possible and/or likely that you will not get all of the transfer events. This is due to limits on the number of results that can be returned, and depends on the RPC provider that is used (e.g. alchemy / Infura, etc.) Some will limit to a fixed number of events, others will have a hard limit on the history (you need an archive node provider to go back beyond a certain point).

Note that in many cases when building a dapp frontend it makes sense to allow the user to select and use their own provider for the requests made through your frontend - in this case you will not know how many results they might get, nor whether it is would be possible to get all events. You can of course configure a separate ethers provider with an API key paid for by AirSwap that would have sufficient capacity, but note that this would likely be very expensive if nothing was cached centrally - every time a user browses a collection they will query the transfer history from the block that the contract is deployed. It will also be slow.

RPC requests from cloud providers like infura charge for usage, and/or have a rate limit. Large log queries are generally the operations that they limit the most because of their potential computational intensity.

Having written all this, I have convinced myself that a best approach is to assume some regularity of the ids in the contract (which is _not_ a requirement, but is extremely common).

Basically you can check `totalSupply` and if that is e.g. 10k, you can assume the IDs are 1 - 10k. If this was incorrect we could have a way that users can manually enter an ID (in an unbounded range) and view that NFT. There will also be cases where e.g. 1-100 don't exist, or some have been burned etc - this can be handled when fetching the metadata and current owner, and could also be used as the event to flag a more in-depth log trawl as described above and in the stack overflow answer.

The other alternative is to maintain some kind of central cache or index for each collection, so the work can be done once when the first user visits that page and stored centrally for other users to benefit from.

**Comment by piersss** [comment]
Ok so we can assume id's are incremental, but if they are not we need to give the admin the option to enter a list of id's manually in the .env file. Then we have a list of tokens based on id that we can show to users. Not sorted by date or price yet though.

@gpxl-dev right now the .env file looks like this: https://github.com/airswap/airswap-marketplace/blob/main/.env.example

**Comment by gpxl-dev** [comment]
@piersss yeah I think this is a good idea to start with. Ofc worth implementing some shorthands so you can do `collection-ids=100-1000, 2500-10000` instead of typing every one out, but I’m sure that was implied & obvious 😅

**Comment by filipmanole** [comment]
Saw the comments a little late. I have a draft implementation for fetching the data: https://github.com/airswap/airswap-marketplace/pull/30
